Fixes and Expansions to modep-ctrl.py

whofferbert opened the pull request:
I added a number of new definitions to the modep-ctrl.py, because when I started trying to use it to change pedalboards, I noticed some problems, and things that could be easily improved.

The first thing I noticed, was it seemed that only some built in pedalboards were returned by the "list" command, which was getting it's info from calling 'get_pedalboards'.

```bash
modep@modep:~/modep-btn-scripts $ ./modep-ctrl.py list
/usr/local/modep/.pedalboards/DISTORTION.pedalboard
/usr/local/modep/.pedalboards/ORGANS.pedalboard
/usr/local/modep/.pedalboards/SEQUENCED_DRUMS.pedalboard
/usr/local/modep/.pedalboards/default.pedalboard
modep@modep:~/modep-btn-scripts $ 
```

I looked in to it, and I'm not sure why get_pedalboards was previously referencing "/banks", but it seemed not right to me. After some cross referencing in https://github.com/BlokasLabs/mod-ui/blob/master/mod/webserver.py , it seemed that it was more proper to look at "pedalboard/list" here.
So therefore, I rewrote get_pedalboards to look at the pedalboards/list instead of /banks.
I needed to adjust a number of other calls accordingly.

Also, seeing some possibly useful and not implemented definitions inside, I added a 'current' option to the main case statement at the end, to simply return the currently used pedalboard:

```bash
modep@modep:~/modep-btn-scripts $ ./modep-ctrl.py current
/usr/local/modep/.pedalboards/DEFAULT.pedalboard
modep@modep:~/modep-btn-scripts $ 
```

The last bit is the biggest thing I was after. I also added a few definitions to make it easier to select pedalboards by their saved names, which works with the short name, full(bundle) name, and return errors as appropriate:

```bash
modep@modep:~/modep-btn-scripts $ ./modep-ctrl.py list
/usr/local/modep/.pedalboards/DEFAULT.pedalboard
/usr/local/modep/.pedalboards/DISTORTION.pedalboard
/usr/local/modep/.pedalboards/GUITAR_PB1.pedalboard
/usr/local/modep/.pedalboards/GUITAR_PB1_NOMID-74428.pedalboard
/usr/local/modep/.pedalboards/GUITAR_PB1_NOMID.pedalboard
/usr/local/modep/.pedalboards/HEXA_SYNTHS.pedalboard
/usr/local/modep/.pedalboards/ORGANS.pedalboard
/usr/local/modep/.pedalboards/SEQUENCED_DRUMS.pedalboard
/usr/local/modep/.pedalboards/default.pedalboard
/usr/local/modep/.pedalboards/figgis.pedalboard
/usr/local/modep/.pedalboards/shutdown_button.pedalboard
modep@modep:~/modep-btn-scripts $ 
modep@modep:~/modep-btn-scripts $ ./modep-ctrl.py load-board HEXA_SYNTHS
Switching /usr/local/modep/.pedalboards/DEFAULT.pedalboard -> /usr/local/modep/.pedalboards/HEXA_SYNTHS.pedalboard
modep@modep:~/modep-btn-scripts $ 
modep@modep:~/modep-btn-scripts $ ./modep-ctrl.py load-board pizza
No board found for pizza!
modep@modep:~/modep-btn-scripts $ 
modep@modep:~/modep-btn-scripts $ ./modep-ctrl.py load-board /usr/local/modep/.pedalboards/DEFAULT.pedalboard
Switching /usr/local/modep/.pedalboards/HEXA_SYNTHS.pedalboard -> /usr/local/modep/.pedalboards/DEFAULT.pedalboard
modep@modep:~/modep-btn-scripts $ 
```

Let me know if there are any troubles with these changes?

whofferbert committed:
added two functions, current and load-pedalboard

whofferbert committed:
cleaning up comments

whofferbert committed:
made some new method definitions for names and indexes, clarified code

whofferbert committed:
added some more method defs to clean up the main case statement

whofferbert commented:
I should also note that all the other functions seem to behave as well, after my adjustments:

```bash
modep@modep:~/modep-btn-scripts $ ./modep-ctrl.py prev
Switching /usr/local/modep/.pedalboards/DEFAULT.pedalboard -> /usr/local/modep/.pedalboards/shutdown_button.pedalboard
modep@modep:~/modep-btn-scripts $ ./modep-ctrl.py next
Switching /usr/local/modep/.pedalboards/shutdown_button.pedalboard -> /usr/local/modep/.pedalboards/DEFAULT.pedalboard
modep@modep:~/modep-btn-scripts $ vim modep-ctrl.py 
modep@modep:~/modep-btn-scripts $ ./modep-ctrl.py bypass
Storing /usr/local/modep/.pedalboards/DEFAULT.pedalboard, switching to default board.
modep@modep:~/modep-btn-scripts $ ./modep-ctrl.py bypass
Restoring /usr/local/modep/.pedalboards/DEFAULT.pedalboard board from default board.
modep@modep:~/modep-btn-scripts $ ./modep-ctrl.py prev
Switching /usr/local/modep/.pedalboards/DEFAULT.pedalboard -> /usr/local/modep/.pedalboards/shutdown_button.pedalboard
modep@modep:~/modep-btn-scripts $ ./modep-ctrl.py next
Switching /usr/local/modep/.pedalboards/shutdown_button.pedalboard -> /usr/local/modep/.pedalboards/DEFAULT.pedalboard
modep@modep:~/modep-btn-scripts $ 
```

gtrainavicius commented:
Hey, as quickly described here: https://community.blokas.io/t/midi-command-server/1001/4, loading from banks was done on purpose. So in order to work with data from /pedalboards, some command line switch should be added (like --all or --pedalboards), this behavior change would break the Pisound button integration in MODEP.

Other changes seem fine. 👍

whofferbert committed:
switched index and prev/next back to bank addressing, added new switches

whofferbert committed:
cleaning up comments

whofferbert commented:
I updated the code with this new information (and I'm surprised I failed to notice banks until now!)

Armed with this new information, I made the list and next/prev switched act on the default bank.

You can also, optionally supply a bank number to 'list' to get the boards in that bank.

I also included a switch to list bank info, and another to list all boards:

```bash
modep@modep:~/modep-btn-scripts $ ./modep-ctrl.py list
/usr/local/modep/.pedalboards/default.pedalboard
/usr/local/modep/.pedalboards/DISTORTION.pedalboard
/usr/local/modep/.pedalboards/SEQUENCED_DRUMS.pedalboard
/usr/local/modep/.pedalboards/DEFAULT.pedalboard
/usr/local/modep/.pedalboards/ORGANS.pedalboard
modep@modep:~/modep-btn-scripts $ ./modep-ctrl.py list-banks
Bank 0: The Button
	Default
	DISTORTION
	SEQUENCED_DRUMS
	DEFAULT
	ORGANS
Bank 1: test_bank
	DEFAULT
	GUITAR-PB1-NOMIDI
	GUITAR-PB1
Bank 2: test_TEW
	HEXA_SYNTHS
	GUITAR-PB1-NOMIDI
modep@modep:~/modep-btn-scripts $ ./modep-ctrl.py list 2
/usr/local/modep/.pedalboards/HEXA_SYNTHS.pedalboard
/usr/local/modep/.pedalboards/GUITAR_PB1_NOMID.pedalboard
modep@modep:~/modep-btn-scripts $ 
modep@modep:~/modep-btn-scripts $ 
modep@modep:~/modep-btn-scripts $ ./modep-ctrl.py list-boards
/usr/local/modep/.pedalboards/DEFAULT.pedalboard
/usr/local/modep/.pedalboards/DISTORTION.pedalboard
/usr/local/modep/.pedalboards/GUITAR_PB1.pedalboard
/usr/local/modep/.pedalboards/GUITAR_PB1_NOMID-74428.pedalboard
/usr/local/modep/.pedalboards/GUITAR_PB1_NOMID.pedalboard
/usr/local/modep/.pedalboards/HEXA_SYNTHS.pedalboard
/usr/local/modep/.pedalboards/ORGANS.pedalboard
/usr/local/modep/.pedalboards/SEQUENCED_DRUMS.pedalboard
/usr/local/modep/.pedalboards/default.pedalboard
/usr/local/modep/.pedalboards/figgis.pedalboard
/usr/local/modep/.pedalboards/shutdown_button.pedalboard
modep@modep:~/modep-btn-scripts $ 
modep@modep:~/modep-btn-scripts $ ./modep-ctrl.py index 4
Switching /usr/local/modep/.pedalboards/HEXA_SYNTHS.pedalboard -> /usr/local/modep/.pedalboards/ORGANS.pedalboard
modep@modep:~/modep-btn-scripts $ ./modep-ctrl.py index 5
Index 5 is out of range!
modep@modep:~/modep-btn-scripts $ ./modep-ctrl.py load-board HEXA_SYNTHS
Switching /usr/local/modep/.pedalboards/ORGANS.pedalboard -> /usr/local/modep/.pedalboards/HEXA_SYNTHS.pedalboard
modep@modep:~/modep-btn-scripts $
```

Notably, I added a print statement to the load_index definition, so it'll tell you what is being switched to. I don't imagine that should break anything, but it's different than it was.

Let me know if this works for you!

whofferbert commented:
Ah, I realize that if you switch to a pedalboard that's outside of bank 0, and then use the next or previous functions, it won't work reliably. I'll poke that soon and get it sorted out.

whofferbert commented:
Actually, upon further inspection, get_current_pedalboard_index returns -1 if it can't find the board name in the current board list, which then behaves sanely, always returning you to the first board in the default bank in that scenario.

I suppose that, if there were a method to check what bank we are currently using, then we could load the board index of that bank, and use next/previous reliably within banks that way... but I didn't see any way to do that with a cursory glance through the mod/webserver.py.

Barring the behavior I was just describing, I think the changes here are sane.

gtrainavicius commented on modep-ctrl.py:
```diff
@@ -104,8 +118,12 @@ def load_prev():
-def load_index(index):
-	boards = get_pedalboards(0)
+def load_index(index, load_all):
+	if load_all == 0:
```
Boolean True/False would be preferable to 0 / 1 values for load_all

gtrainavicius commented:
I have taken a look at the updated code, and it looks good, just added a small comment to the code.

I'll give it a try tomorrow with MODEP and the Pisound Button and App, and will merge if everything works fine. :)

whofferbert committed:
replaces 0/1 values in favor of False/True

whofferbert commented:
I come from a mostly bash and perl 5 background; still learning the pythonic way of doing things.

Thanks for the advice, it does make it more clearly understandable with True/False.

whofferbert committed:
removed current bank? comments

whofferbert committed:
switched test order so the test reads if True rather than if False

gtrainavicius commented:
Ok, tried it out, working well! Is this ready to be merged?

whofferbert commented:
As far as I know, yes. I think I'm happy with it if you are

On Fri, Feb 8, 2019, 03:44 gtrainavicius <notifications@github.com wrote:

> Ok, tried it out, working well! Is this ready to be merged?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/BlokasLabs/modep-btn-scripts/pull/1#issuecomment-461730639>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/Abiv1sHcqnZA4pyX-iRGMAS5tLZIK6aIks5vLTjigaJpZM4aijQo>
> .
>
